lib/internal/shutil.js

```diff
@@ -0,0 +1,294 @@
+const fs = require('fs');
+const { join } = require('path');
+const { validatePath } = require('internal/fs/utils');
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_CALLBACK
+  }
+} = require('internal/errors');
+
+const isWindows = process.platform === 'win32';
+const _0666 = parseInt('666', 8);
+
+// for EMFILE handling
+let timeout = 0;
+
+function rmtree(path, options, callback) {
+  if (typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
+
+  validatePath(path);
+  validateCallback(callback);
+
+  options = getOptions(options);
+
+  let busyTries = 0;
+
+  rmtree_(path, options, function CB(er) {
+    if (er) {
+      if (
+        (er.code === 'EBUSY' || er.code === 'ENOTEMPTY' || er.code === 'EPERM') &&
+        busyTries < options.maxBusyTries
+      ) {
+        busyTries++;
+        let time = busyTries * 100;
+        // try again, with the same exact callback as this one.
+        return setTimeout(function() {
+          rmtree_(path, options, CB);
+        }, time);
+      }
+
+      // this one won't happen if graceful-fs is used.
+      if (er.code === 'EMFILE' && timeout < options.emfileWait) {
+        return setTimeout(function() {
+          rmtree_(path, options, CB);
+        }, timeout++);
+      }
+
+      // already gone
+      if (er.code === 'ENOENT') er = null;
+
+      callback(er);
+    }
+
+    timeout = 0;
+    callback();
+  });
+}
+
+// Two possible strategies.
+// 1. Assume it's a file.  unlink it, then do the dir stuff on EPERM or EISDIR
+// 2. Assume it's a directory.  readdir, then do the file stuff on ENOTDIR
+//
+// Both result in an extra syscall when you guess wrong.  However, there
+// are likely far more normal files in the world than directories.  This
+// is based on the assumption that a the average number of files per
+// directory is >= 1.
+//
+// If anyone ever complains about this, then I guess the strategy could
+// be made configurable somehow.  But until then, YAGNI.
+function rmtree_(path, options, callback) {
+  validatePath(path);
+  validateOptions(options);
```

Comment:
This looks pretty close to the original `rimraf` logic, we should keep the original license header if we're bringing over @isaacs' code.

Reply:
Agreed. We discussed that in our last meeting.